http/server.lua

```diff
@@ -0,0 +1,297 @@
+end
+
+local function httpd_router(self)
+    return self.__v2_server:router()
+end
+
+local function __set_v1_handler(self, handler)
+    return self.__v1_server_adapter:set_handler(handler)
+end
+
+local function server_route(self, opts, handler)
+    self.__v1_server_adapter:route(opts, handler)
+    return self
+end
+
+local function server_match(self, method, route)
+    return self.__v1_server_adapter:match(method, route)
+end
+
+local function server_helper(self, name, handler)
+    self.__v1_server_adapter:helper(name, handler)
+    return self
+end
+
+local function server_hook(self, name, handler)
+    self.__v1_server_adapter:hook(name, handler)
+    return self
+end
+
+local function server_url_for(self, name, args, query)
+    return self.__v1_server_adapter:url_for(name, args, query)
+end
+
+local server_fields_set = {
+    host        = true,
+    port        = true,
+    tcp_server  = true,
+    is_run      = true,
+}
+
+local router_fields_set = {
+    routes      = true,
+    iroutes     = true,
+    helpers     = true,
+    hooks       = true,
+    cache       = true,
+}
+
+local v1_server_options_set = {
+    log_requests        = true,
+    log_errors          = true,
+    display_errors      = true,
+}
+
+local v1_router_options_set = {
+    max_header_size     = true,
+    header_timeout      = true,
+    app_dir             = true,
+    charset             = true,
+    cache_templates     = true,
+    cache_controllers   = true,
+    cache_static        = true,
+}
+
+local function is_v1_only_option(option_name)
+    return v1_router_options_set[option_name] ~= nil or option_name == 'handler'
+end
+
+local function is_v2_only_option(option_name)
+    return option_name == 'router'
+end
+
+local function get_router_options_for_v1(options)
+    local result = {}
+    for option_name, _ in pairs(v1_router_options_set) do
+        result[option_name] = options[option_name]
+    end
+    return result
+end
+
+local function get_v2_server_options(options)
+    return {
+        router              = options.router,
+        log_requests        = options.log_requests,
+        log_errors          = options.log_errors,
+        display_errors      = options.display_errors,
+    }
+end
+
+local function httpd_start(self)
+    if self.__api_version == API_VERSIONS.UNKNOWN then
+        local router = v2_router.new(get_router_options_for_v1(self.options))
+        self.__api_version = API_VERSIONS.V1
+        self.__v2_server:set_router(router)
+    end
+    return self.__v2_server:start()
+end
+
+local function v1_method_decorator(method, method_name)
+    return function(self, ...)
+        if self.__api_version == API_VERSIONS.V1 then
+            return method(self, ...)
+        elseif self.__api_version == API_VERSIONS.V2 then
+            error(
+                ('":%s" method does not supported. Use http-v2 api https://github.com/tarantool/http/tree/master.'):
+                format(method_name)
+            )
```

Comment:
Is the check http server instance wide? So, different http server instances may be used according to different API versions, right? Is it fit good to the case, when cartridge uses one API version, but a user's application uses another one?

What is the reason of blocking such usage? It would not work appropriately?

Reply:
It is a check only for one http_server instance, not for all.
Yes, different instances could be used according to different API versions.
Yes, I think it fits good to the case with user and cartridge application servers. We may ask @rosik about it.

Both APIs have documentation and covered by tests. But their simultaneous usage not documented and not tests. So I think there are cases when simultaneous usage can lead to bugs or strange not documented behaviour. Error on such cases prevents it. Also I don't think that simultaneous usage is good practice. As for me it would not give any advantages and confuses a code with a lot of method calls, which do the same thing.

Reply:
Thanks for the clarification! It seems meaningful.

Reply:
@rosik said that users usually add application specific routes into the cartridge's http server.